packages/nitro-protocol/contracts/ninja-nitro/readme.md

```diff
@@ -0,0 +1,97 @@
+
+### Deposits
+
+In Mono-Nitro, depositing requires calling a method on each `AssetHolder`.
+In Ninja-Nitro, depositing can be achieved simply by sending funds or tokens to the address of the `Proxy`, and this can even be done before the `Proxy` is created/deployed.
+
+### Adjudicator status
+
+Ninja-Nitro was implemented by modifying a copy of Mono-Nitro. This means that certain quirks remain in the implementation which would probably not exist had it been written from scratch. These quirks may or may not be removed in future.
+
+One such quirk is as follows. In Mono-Nitro, the hash of the status of a channel (which includes the full outcome, the address of any challenger, the time when the channel will finalize, etc) is stored in a mapping, keyed by the `channelId`.
+
+In Ninja-Nitro, exactly the same is true: only each proxy has its own mapping. The reason that this is quirky is that each of these mappings is only ever going to have a single entry, because each proxy is in a 1-1 relation to a `channelId`. The reason that this quirk is not actually very troublesome are:
+
+- the gas cost are not dramatically higher than storing a single hash without using a mapping (it may in fact cost precisely the same amount). See https://docs.soliditylang.org/en/v0.8.3/types.html?highlight=mapping#mapping-types
+- we are able to reuse `ForceMove` (in the sense of `SingleChannelAdjudicator is ForceMove`) in its entirety without modifying it at all
+
+> A future optimization might be to replace the mapping with a regular, singleton storage variable. This would require a small but breaking change to `ForceMove.sol`.
+
+### External destinations
+
+In Mono-Nitro, funds for channels are tracked _internally_, while funds for any Ethereum address (e.g. an externally owned account) are typically paid out of the system when defunding a channel. Padding an Ethereum address with enough zeros results in a 32 byte number which we call an external destination. Channel ids are known as internal destinations.
+
+In Ninja-Nitro, funds are always tracked externally. This is possible since (up to hash collisions) each channel now has a unique Ethereum address. Funds are therefore always paid _out_ of a `Proxy` contract after a `transfer` or `claim`. There is therefore no such thing as an internal destination in Ninja-Nitro. Everything is external.
+
+To maintain compatibility with the off-chain code we already have for Mono-Nitro, the Ninja-Nitro implementation should convert destinations to addresses in the following way:
+
+- if the destination is a padded Ethereum address, slice of the Ethereum address and use that
+- if the destination is not a padded Ethereum address, compute the create2 address of that channel and use that
+
+> TODO it does not currently do this and our test suite is not adequate enough to tell us!
```

Comment:
I will create a New Issue for this.

Reply:
Does our test suite also not detect this for the current nitro?

Reply:
Our current test is OK because it has cases that allocate to a channel as well as to an external destination. The ninja-nitro test suite flipped everything to being an external destination, and so was  bypassing an important code path. Fixing that over on #3492